### ISSUE TITLE
[feature request] Support ignore block of code with noqa

### ISSUE THREAD

**Issue by woile** [charliermarsh/ruff#3711]
First of all, thank you so much for this tool, it's fantastic 🚀 ! 

I would like to request the possibility to support ignoring a block of code. To my surprise, this is a feature some people have needed as well from flake8, [see SO question](https://stackoverflow.com/questions/64428794/flake8-disable-linter-only-for-a-block-of-code), though it is not supported.

In my current use case, I have a bunch of templates, that are to long, and it becomes cumbersome to add `noqa` to all of them.

I was wondering if adding something like this would be possible (maybe with a different syntax):

```python
# ruff: noqa: on: E501
TEMPLATE = """
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque sit amet ligula magna. Nulla ${var1} vehicula leo, eget tincidunt leo pharetra in. Ut egestas felis et vehicula commodo. Aliquam ${var2} placerat tortor sed magna faucibus facilisis ut et mauris. Maecenas vitae sem augue. Cras egestas felis nulla, id porta nisl ${var3} vel
"""
# ruff: noqa: off: E501
```

Thanks!

**Comment by charliermarsh** [comment]
I'd really like to support something like this. But we may want to couple it with a more holistic redesign of suppression comments. \cc @MichaReiser

**Comment by charliermarsh** [comment]
And thank you for the kind words :)

**Comment by whinee** [comment]
That the program ignores the lines under that comment.

A markdown linter named [markdownlint](https://github.com/DavidAnson/markdownlint) has [that option](https://github.com/DavidAnson/markdownlint#configuration) and if I am not mistaken, if there is no `enable` comment down the line, then the program just ignores all of it.

This program could also have an option to warm the user when an off flag/comment is unterminated (has no corresponding in switch) cuz disabling linting for the rest of the file is not recommended???? I dont knowwwwwwww

**Comment by evanrittenhouse** [comment]
I'm pretty interested in this one. Obviously, pretty new to the project, but I think that the design proposed above makes sense. Not sure about any overarching design changes - @MichaReiser saw you were mentioned above, are we good to move forward on this design? Do we want to support ignoring multiple rules at once?

**Comment by MichaReiser** [comment]
> @MichaReiser saw you were mentioned above, are we good to move forward on this design? Do we want to support ignoring multiple rules at once?

Technically: I recommend waiting till after #3931 lands because it significantly changes how we handle noqa comments internally. 

Regarding the design. I haven't had much time to look into suppression comments, I got sidetracked by figuring out how a setting format could look like in a world where ruff does formatting and linting (and more?). My personal objective for suppression comments is that we should try to keep it simple. I really like Rust's approach where there's a single way to enable or disable lints: Using `allow` and `deny`. The scope is defined by where the attributes are put: In a module -> applies for the whole module, at the top of a function -> applies for the function, at the top of an expression -> applies to the expression. Having a single format keeps things simple and easy to remember. I've had to google ESLint's suppression comment format so many time in the past, got it wrong because I forgot the `on` comment after disabling a rule for a section, or accidentally deleted the `on` comment in a refactor :hand_over_mouth: .

I'm not proposing that we need to implement the same semantics but I think its worthwhile to list the use cases with examples on how the new suppression comments should work. Do we support whole file suppression? Do we want to warn about missing `on` comments after an `off`. etc.

**Comment by evanrittenhouse** [comment]
That makes sense. I'll try to come up with a design - since #1054 asks for block/file-scoped `noqa` directives, closing this issue should also close that one (file-scoped `noqa`s are already implemented)

**Comment by evanrittenhouse** [comment]
We can build on the syntax introduced in #2978 to allow code blocks by enabling/disabling `noqa` selectors

Note that all approaches can be easily extended to include multiple rules by following the typical convention of `<rule1>,<rule2>,...`. 

We can also easily extend both approaches to add _enabling_ on a per-block basis in the future, if we need to. We could build in logic to parse the first suppression comment (either `disable` or `enable`) and tailor our logic around that.

## Approach 1
```python
z = 1

# ruff: noqa: disable: E741
x = 1
y = 1
# ruff: noqa: enable: E741

# ruff: noqa: disable: E741
def somePoorlyDefinedVariable():
	print("this won't throw")
	z = 1
# ruff: noqa: enable: E741
```

This should yield one `AmbiguousVariableName` diagnostic on line 1

### Pros/Cons
1. (+) Easily legible and easy to reason about for the reader/user
2. (+) More easily implemented - simply find the lines between the two tags and ignore all diagnostics in them
3. (-) Clunkier for shorter blocks
4. (-) More verbose than other options. 
5. Also slightly more fragile than other options since you could delete either a `disable` or `enable` without deleting the other, which would prompt a new diagnostic to show up. That could get annoying, though you could make the argument that it also makes the code more robust since you can't accidentally retrigger diagnostics. More of a nit than anything, IMO
6. (-) Doesn't fit the existing ecosystem - more similar to [eslint's](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1) implementation than Pylint's

### FAQ
1. What happens if `E741` isn't enabled for the file being edited and I try to disable it? What if it's enabled and I try to re-enable it?
	1. I think a new `RUF` rule would be appropriate which alerts the user to an unnecessary `noqa` block. We could also of course simply ignore it, but I think that goes against Ruff's philosophy - unnecessary suppresion comments would clutter up the code with no good reason
2. Will this ever conflict with `# ruff: noqa: E741`? 
	1. Good question. I think not, since that line is already the trigger for the "disable-everywhere" function. That directive could continue to function as-is, and blocks would be delineated by `disable/enable` tags.  This does lead to the issue of multiple syntaxes, though, which is something I'd like to avoid.
3. What if we disable without ever re-enabling/enable without ever disabling?
	1. In my opinion, it's more in line with Ruff's philosophy to raise a new diagnostic mentioning that you've opened a suppression block without closing it.


## Approach 2
```python
z = 1

# ruff: noqa:< disable:> E741
x = 1
# ruff: noqa:< disable:> E741
y = 1

# ruff: noqa:< disable:> E741
def somePoorlyDefinedVariable():
	print("this won't throw")
	z = 1

       # ruff: noqa: enable: E741
      a = 1 # this will throw
```

This should only throw on line 2 - the idea behind this approach is that the `disable` functions on the AST node directly beneath it. Directives should be prioritized by precision - i.e., `a  = 1` should throw because the rule is enabled in a scope which is narrower than the scope which disables it.

 This is pretty similar to Pylint's current [block comment implementation](https://pylint.readthedocs.io/en/latest/user_guide/messages/message_control.html#block-disables). If we went with this approach, it may be worth considering whether we should simply copy their ruleset but with a different directive comment.

I put `< disable:>` in braces to delineate the fact that we _could_ treat it as optional. I don't think that's a great idea though; if we did, the current `# ruff: noqa` directive would need to be refactored to act in a per-block fashion and we'd have to figure out how to signify per-file suppressions. Long story short, if we went with this approach I'd strongly recommend having it be:
`# ruff: noqa: disable: E741`



### Pros/Cons
1. (+) Quicker to write for users
2. (+) Similarity to Pylint will make it more familiar to users, making it easier to plug and play
3. (-) As far as I can tell, implementation for this may be more difficult than Approach 1
4. (-) Can be a bit unintuitive, especially in the case of lines 2/3 - it's somewhat tough to tell off the bat that we only want to suppress on line 2
7. (-) More fragile - if you delete one line accidentally, you could disable/enable rules easily

### FAQ
1. What happens if `E741` isn't enabled for the file being edited and I try to disable it? What if it's enabled and I try to re-enable it?
    1. Similar to Approach 1, in this case I think that we should throw a new diagnostic.
1. Will this ever conflict with `# ruff: noqa: E741`? 
    1.  I think it _should_. To preserve Python's "one way of doing something" ethos, under this approach we may want to consider changing file-wide suppression so it's only activated by putting `# ruff: noqa: disable: E741` at the top of the file. This is in contrast to the current way of doing this: `# ruff: noqa: E741` will disable everywhere, regardless of where the file is. I'd like to avoid the issue of having one syntax for 
1. What if we disable without ever re-enabling/enable without ever disabling?
    1. That wouldn't be possible with this approach. The suppression block would simply go out of scope, then we could continue to read diagnostics as normal

*Rejected*: I (very briefly) considered a `pyproject.toml`configuration option, but that's pretty infeasible given how often code changes. While it could clean up the code, it'd be more difficult to reason about ("why are these random diagnostics showing up? I haven't disabled them in my code anywhere? Ah they're in `pyproject.toml` for some reason")  and changing them whenever code lines change would be...tedious, to say the least.

## Verdict
I'm leaning more towards Approach 2 since it more closely mirrors the existing Python ecosystem. I remember some issue mentioning that a big component of Ruff's growth was the ease of dropping it into an existing project. Approach 2's cognitive overhead is lower than Approach 1's as it would act similarly to Pylint's block disabling. On the other hand, Approach 1 is more robust - it's harder to inadvertently change things. The biggest factor for me is the issue of multiple syntaxes - Approach 2 would maintain a consistent syntax/API that, once understood, is quite simply to understand.


CC @charliermarsh @MichaReiser let me know what you think! I think either way implementation will wait until #3931 as mentioned above, so no rush. Open to any feedback that you may have!

**Comment by tekumara** [comment]
FYI - duplicate of #1289

**Comment by adam-grant-hendry** [comment]
FWIW, I believe suppression for any scope as in [`pylint` block disables](https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html#block-disables) in addition to section-level suppression (i.e. turning on/off linting) would be wise.

Without knowing the internals, I imagine the former would require reparsing the AST, and thus a significant rewrite. However, 

- it would match existing `pylint` behavior
- I’m slowly seeing several different feature requests for suppression in different scopes (global level, function level, etc.).

Rather than rewrite the source for every kind of scope feature request, this would be done in one change and would likely cover 99% of use cases for most users.

**Comment by flbraun** [comment]
I'd like to emphasize the importance of this feature for ruff's usability. There are some situations where you _need_ more sophisticated rule control, for example this:

```python
def get_all(cursor, table_name):
    return cursor.execute(
        '''
        SELECT *
        FROM {table_name};
        '''.format(table_name=table_name)
    )
```

ruff rightfully complains that this code contains a possible sql injection (S608). We know that the value of `table_name` is safe, so we want to disable S608 for that statement.
We can't disable the rule with `# noqa` because the affected line opens a multi-line string, so the only other option we have is to disable it for the entire file. Which is bad, because that would also disable it for any other sql statement in that file.
The only correct solution would be to wrap the entire sql string in an off-on-block or utilizing a `disable-next` pragma.

**Comment by ThiefMaster** [comment]
Why don't you pass it as a param like you should? Sorry, but this is a perfectly true positive, even if in THIS particular case it's safe - until someone decides to make `cond` come from let's say user-provided input. Boom, SQL injection!

**Comment by m-u-xyz** [comment]
Passing it as a param works in this simple case, but when you build a whole query programmatically …?

**Comment by flbraun** [comment]
> Why don't you pass it as a param like you should? Sorry, but this is a perfectly true positive, even if in THIS particular case it's safe - until someone decides to make `cond` come from let's say user-provided input. Boom, SQL injection!

Sorry, my example was chosen very poorly, I've replaced it with a less frivolous one where you can't get away with parameter-passing.

**Comment by MichaReiser** [comment]
@flbraun I believe your example might be worth raising a separate issue. I would have expected that one of those work, but none of them do.

```python
def get_all(cursor, table_name):
    return cursor.execute(
        f'''
        SELECT *
        FROM {table_name};
        '''  # noqa: S608
    ) 

def get_all(cursor, table_name):
    return cursor.execute(
        (
            """
        SELECT *
        FROM {table_name};
        """  # noqa: S608
        ).format(table_name=table_name)
    )
```

**Comment by kaddkaka** [comment]
Is there any ongoing work on this issue?

**Comment by Xavier-Do** [comment]
We are also needing for this feature

Our use case is mainly in tests, we like to disable whitespaces rules the time of an assertions, to present a list/tuple/dict as a table, with aligned rows. 

```python
        self.assertEquals(
            # pylint: disable=bad-whitespace
            report,
            [
                ['Header',              400.00,      1600.00,          100.00],
                ['Other header',         00.00,       600.00,        10000.00],
                ['Some Header',         140.00,         0.00,         1000.00],
                ['',                   1400.00,       600.00,          100.00],
                ['Finally',           10400.00,       600.00,          100.00],
            ]
        )
```

**Comment by Suor** [comment]
Maybe instead of `# ruff: noqa: on` and `... off` simply:
```python
# ruff: off
...
# ruff: on

# ruff: off: T201
print()
if ...:
    print("...", a, b, c)
...
# ruff: on
```

And since we have already `# ruff: noqa ...` to do it file level I would warn about non-paired `off` with a suggestion of either close it or replace with `ruff: noqa: whatever`.

**Comment by Pierre-Sassoulas** [comment]
I think there's a performance aspect to consider. In pylint the fine grained message control sometime prevented us from bypassing costly processing because checking if a message is disabled or not (on a directory, file, scope, line, node, or token) is costly in itself.

**Comment by kaddkaka** [comment]
> I stumbled upon another use case where this would be useful:
> 
> A line of code causes one error for mypy and one for bandit. Both support inline-comments for ignoring it, but they don't support ignoring code blocks. I can only ignore the error for one of them and have to ignore the whole file for another (not nice). If I used Ruff (with this feature) as a replacement for Bandit, I could ignore the block of code (1 line) and also add the inline ignore comment for mypy. So it would improve compatibility with mypy (they also have an [ongoing thread](https://github.com/python/mypy/issues/6948) about implementing this) and other linters without this feature.

Why not just put two inline comments on the same line?

**Comment by AntiSol** [comment]
Just to chime in with my $0.02, I also think a good linter should have the ability to ignore a block

> What's your expectation if a file only contains an `off` comment without a corresponding `on` comment later in the file?

It should ignore the rest of the file (but turn itself back on at the end of the file)

**Comment by BobVitorBob** [comment]
Another use case for this.

I wanted to leave a comment explaining why I'm turning another rule off, which includes a snippet of a minimal example for a pandas warning. Since it's commented python code, it lights up with `ERA001`. In order to suppress this, I have to do `# noqa: ERA001` for each line, which is unnecessary imo.

```
# import numpy as np                                                                # noqa: ERA001
# import pandas as pd                                                               # noqa: ERA001
# df = pd.DataFrame({"col1": [1, 2, 3, "", 5]}, dtype=object)                       # noqa: ERA001
# print(df)                                                                         # noqa: ERA001
# print(df.dtypes)                                                                  # noqa: ERA001
# df2 = df.replace("", 0)                                                           # noqa: ERA001
# print(df2["col1"])                                                                # noqa: ERA001
# df2.iloc[0, 0] = "a"                                                              # noqa: ERA001
# print(df2["col1"])                                                                # noqa: ERA001
```

**Comment by AntiSol** [comment]
Use case: 

```python

from a_module import (          # noqa: I001
   # functions                                               
   do_something,                # noqa: I001
   do_something_else,           # noqa: I001

   # variables
   some_module_config_var,      # noqa: I001
   some_other_module_config_var,# noqa: I001
   fifty_more_of_these,         # fifty more noqa: I001

)

```

because I wasn't doing that awfulness, and there's no way to disable this rule for a block of code, now my code has no import sorting.

**Comment by dhruvmanila** [comment]
@AntiSol You shouldn't require to specify `noqa` comment for each line, just one should suffice: https://play.ruff.rs/4c44cfef-9821-4cd6-98fd-6e3187228e68

**Comment by AntiSol** [comment]
@dhruvmanila I actually tried that before posting. 

I don't think you're running the checker there? just the formatter? Isort rules are applied by the checker, not the formatter. I don't see any way to run the checker / isort rules there? 

You'll note that removing the noqa comment from your example [has no effect on the formatter output](https://play.ruff.rs/4c44cfef-9821-4cd6-98fd-6e3187228e68?secondary=Format)
![image](https://github.com/user-attachments/assets/e924e051-f8d6-4726-9993-ea732cc99498)
Maybe I'm missing something?

**Comment by dhruvmanila** [comment]
Correct me if I'm wrong but I'm assuming that you're trying to ignore the names in that import statement from being sorted, right?


https://github.com/user-attachments/assets/74627bdd-f4a6-4843-9e81-d7709b525ee7

The video shows how to use code actions to apply a fix from the rules in the playground.

**Comment by AntiSol** [comment]
that is what I'm trying to do, and that is what I tried before posting. Maybe I got it wrong somehow. I'll try again when I have time.

Aha, thanks... but I assume you're right-clicking in that video? I don't see that context menu, just the regular one
![image](https://github.com/user-attachments/assets/26c3859d-ff76-4fcd-83ed-44fa35092314)
Which means that there appears to be no way to run the checker in the playground? Might be worth a separate bug report I guess

**Comment by dhruvmanila** [comment]
> Aha, thanks... but I assume you're right-clicking in that video? I don't see that context menu, just the regular one

No, I'm not right-clicking in the video. I'm just hovering over the squiggly lines and the pop-up will appear and then I'm clicking on the "Quick Fix" action which displays the code actions and then you can select any that's available.

> Which means that there appears to be no way to run the checker in the playground? Might be worth a separate bug report I guess

The checker runs automatically in the playground every time the content changes, so there's no need to run it manually. It's not a bug, it's a feature :)

**Comment by AntiSol** [comment]
@dhruvmanila right you are on all counts! And I was also able to confirm that this does indeed solve my issue, so you can just ignore all my comments on this thread. Thank you muchly! :)

**Comment by MilkClouds** [comment]
looking forward to this feature..
![image](https://github.com/user-attachments/assets/27df3f29-9b26-4b2a-b1f9-a318a0b1c208)

**Comment by Daraan** [comment]
Any updates on this?

**Comment by j-adamczyk** [comment]
Any news on this? It would be super useful in chemoinformatics, where I have lists of very long strings (something like regular expressions), and they cannot be made shorter, causing E501 (line length) error. I would like to wrap the entire multiline list in a block to ignore this specific error.

**Comment by berislavlopac-wpp** [comment]
I just discovered that `ruff` apparently ignores `#noqa` comments in doctests! I have a doctest with way too long lines, and it reports `E501` despite the presence of `#noqa` comments in each line. 🤔

**Comment by tigerhawkvok** [comment]
I think the scope-level implementation is best as I have essentially the exact same reasoning as https://github.com/astral-sh/ruff/issues/3711#issuecomment-1512712449

IMO, I'd:

1. Change the top-of-the-file style (`#ruff: noqa: FOO01`) to internally be a scope-level disable, and just add the understanding of other scoping; 
2. add the special command `#ruff: noqa: [code] off` to disable a scope-level `noqa` (eg, re-enable the rule) from that point forward;
3. add a special `RUF` error message for a top-scope `noqa` begun after the first code line that was never disabled

**Comment by Baksalyar** [comment]
Hi! Just wondering — is this still unimplemented, or is it an intentional design decision?
It gets pretty annoying when you need to use something Ruff complains about in some isolated block of code, and the only option is to slap #noqa on every single line. :(

**Comment by Hoernchen** [comment]
Yeah this is turning into a problem everwhere because all kinds of packages need special handling/patching and the only sane way is to just skip E402 completely..

**Comment by MichaReiser** [comment]
This is on my mind and something I want us to fix in the coming months. I just can't promise you a date yet but we didn't forget about it.

**Comment by maltevesper** [comment]
I have not read the entire discussion, but have not found it in the checklist in the issue.
if you do implement it, it might be nice to think about scopes as well: see i.e. https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html

https://github.com/astral-sh/ruff/issues/3711#issuecomment-1487991823
> What's your expectation if a file only contains an `off` comment without a corresponding `on` comment later in the file?

If you have implemented scopes, the answer to your previous comment about how to handle an `off` without a following `on`, on the implementation side might be simply create implicit per-file scopes. Moreover, if you disable multiple lints for a block of code, you do not have to repeat yourself. Rather than typing out the lint numbers in an off comment and a subsequent on comment you simply pop the scope.

**Comment by amyreese** [comment]
For folks interested in testing a new feature in Ruff's preview mode, the 0.14.9 release now supports [block level range suppressions](https://docs.astral.sh/ruff/linter/#block-level) when running with `--preview`:

```py
# ruff: disable[E501]
VALUE_1 = "Lorem ipsum dolor sit amet ..."
VALUE_2 = "Lorem ipsum dolor sit amet ..."
VALUE_3 = "Lorem ipsum dolor sit amet ..."
# ruff: enable[E501]
```

An upcoming release should include new diagnostics for tracking unused or invalid range suppressions, similar to what is currently available for `#noqa`.

Let us know if you have any feedback on the new system.

**Comment by m-u-xyz** [comment]
Question:

> An "enable" comment can only be used to terminate a preceding "disable" comment with identical codes.

Does that mean that

    # ruff: disable[A,B]
    ...
    # ruff: enable[A]
    ...
    # ruff: enable[B]

(or vice versa) won't work?

**Comment by Xavier-Do** [comment]
This is an amazing feature! 🥳 

It looks like it works fine in most cases but there is a specific one supported by pylint that doesn't work here regarding "implicit range"

```python
[
    # ruff: disable[E241]
    {'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
    {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
]
```

In this case the disable is ignored. I would expect to only apply on the content of the list

Another small suggestion would be to support rules names instead of code to improve readability 
```python
def check():
    # ruff: disable[multiple-spaces-after-comma] 
    [
        {'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
        {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
    ]
```

In any case the current version is already a game changer and enough for most needs so thanks a lot for adding this to ruff.

EDIT: I wasn't able to make it work by adding preview in the config, looks like it works only using the command line

**Comment by amyreese** [comment]
> Question:
> 
> > An "enable" comment can only be used to terminate a preceding "disable" comment with identical codes.
> 
> Does that mean that
> 
> ```
> # ruff: disable[A,B]
> ...
> # ruff: enable[A]
> ...
> # ruff: enable[B]
> ```
> 
> (or vice versa) won't work?

Correct. We wanted to keep the implementation simpler, and make it easier to reason about which suppression comments were matched or unmatched and what ranges of code they cover.

**Comment by kaddkaka** [comment]
> > Question:
> > > An "enable" comment can only be used to terminate a preceding "disable" comment with identical codes.
> > 
> > 
> > Does that mean that
> > ```
> > # ruff: disable[A,B]
> > ...
> > # ruff: enable[A]
> > ...
> > # ruff: enable[B]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (or vice versa) won't work?
> 
> Correct. We wanted to keep the implementation simpler, and make it easier to reason about which suppression comments were matched or unmatched and what ranges of code they cover.

So one would ha to write something like:
```py
# ruff: disable[A]
# ruff: disable[B]
...
# ruff: enable[A]
...
# ruff: enable[B]
```

Is this supported?
```py
# ruff: disable[E200]
... 
# ruff: disable[E]
...
# ruff: enable[E]
...
# ruff: enable[E200]
```

**Comment by amyreese** [comment]
> This is an amazing feature! 🥳
> 
> It looks like it works fine in most cases but there is a specific one supported by pylint that doesn't work here regarding "implicit range"
> 
> [
>     # ruff: disable[E241]
>     {'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
>     {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
> ]
> 
> In this case the disable is ignored. I would expect to only apply on the content of the list

I'll try to add a test case for that and see if that's feasible to support. The logic currently depends on Indent/Dedent tokens for tracking blocks/scopes rather than arbitrary indentation on the comments themselves.

> Another small suggestion would be to support rules names instead of code to improve readability
> 
> def check():
>     # ruff: disable[multiple-spaces-after-comma] 
>     [
>         {'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
>         {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
>     ]

I believe this is on our roadmap in general, including for `#noqa` comments. We want to make sure we have consistency across all suppression comment styles.

> EDIT: I wasn't able to make it work by adding preview in the config, looks like it works only using the command line

Can you please open an issue for that with a repro?

> In any case the current version is already a game changer and enough for most needs so thanks a lot for adding this to ruff.

💜

**Comment by amyreese** [comment]
> So one would ha to write something like:
> 
> ruff: disable[A]
> ruff: disable[B]
> ...
> ruff: enable[A]
> ...
> ruff: enable[B]

Yes.

> Is this supported?
> 
> ruff: disable[E200]
> ... 
> ruff: disable[E]
> ...
> ruff: enable[E]
> ...
> ruff: enable[E200]

No, we only support exact code matches, similar to `#noqa`.

**Comment by MichaReiser** [comment]
I think not supporting within expression suppressions is okay. Just suppress the entire statement.

```py
[
    # ruff: disable[E241]
    {'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
    {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
]
```

It otherwise raises questions whether the following is valid or not


```py
[
    {
	    # ruff: disable[E241]
		'line_id': 1,              'balance': 1000.0,  'amount': 2000.0},
	# ruff: enable[E241]
    {'line_id': 2,              'balance': -1000.0, 'amount': -2000.0},
]
```

or even without the enable, how far does the suppression go? Is it also supposed to work in f-strings, what about if I put it in front of a binary expression?


I think this is something where a `ruff: ignore` might be more useful.

**Comment by maltevesper** [comment]
I get false positives for ERA001: The comment is indistinguishable from defining a variable called `ruff` with generic typeannotation `disable[D102]`

```
class Test:
    # ruff: disable[D102]
    def test_method(self):
        pass
```

```
ERA001 Found commented-out code
 --> teste.py:2:5
  |
1 | class Test:
2 |     # ruff: disable[D102]
  |     ^^^^^^^^^^^^^^^^^^^^^
3 |     def test_method(self):
4 |         pass
  |
help: Remove commented-out code
```

Also it would be nice, if an enable is missmatched generated an error. I.e.:
```
class Test:
    # ruff: disable[D102]
    def test_method(self):
        pass

    # ruff: enable[D103]                    <- I would expect to get an error message for this line, since it was not disabled before
```

**Comment by maltevesper** [comment]
I am excited for the new feature and gave it a bit more thought.
I am wondering if it is better to adjust the ERA001 lint or the syntax. However, I find the current syntax compelling, adding a space after the disable does not help (I am not sure why `x : list [int]` parses, but it does) and I did not come up with an alternative syntax I like

**Comment by MichaReiser** [comment]
I think we should change `ERA001` to ignore this known false positives

**Comment by ulgens** [comment]
+1 for excluding this from ERA001 but I missed why we are expecting ERA001 to handle an unimplemented (yet) feature, isn't it a bit early for that?

**Comment by amyreese** [comment]
> +1 for excluding this from ERA001 but I missed why we are expecting ERA001 to handle an unimplemented (yet) feature, isn't it a bit early for that?

The new suppression system is already implemented and available when running Ruff with `--preview`.

**Comment by amyreese** [comment]
> Also it would be nice, if an enable is missmatched generated an error. I.e.:
> 
> ```
> class Test:
>     # ruff: disable[D102]
>     def test_method(self):
>         pass
> 
>     # ruff: enable[D103]                    <- I would expect to get an error message for this line, since it was not disabled before
> ```

That is currently being added in #21908
